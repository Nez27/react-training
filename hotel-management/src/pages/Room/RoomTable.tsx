import { useCallback } from 'react';

// Components
import Menus from '@src/components/Menus';
import Table from '@src/components/Table';
import Message from '@src/components/Message';
import Search from '@src/components/Search';
import SortBy from '@src/components/SortBy';
import OrderBy from '@src/components/OrderBy';
import RoomRow from './RoomRow';

// Types
import { IRoom } from '@src/types/room';

// Constants
import { ORDERBY_OPTIONS, ROOM_PAGE } from '@src/constants/commons';

// Styled
import { StyledOperationTable } from './styled';
import Direction from '@src/commons/styles/Direction';
import Spinner from '@src/commons/styles/Spinner';

// Hooks
import { useRooms } from '@src/hooks/rooms/useRooms';
import Pagination from '@src/components/Pagination';

const RoomTable = () => {
  const columnName = ['Id', 'Name', 'Price', 'Status'];
  const {
    isLoading,
    rooms,
    count
  } = useRooms();

  const renderRoomRow = useCallback(
    (room: IRoom) => <RoomRow room={room} key={room.id} />,
    []
  );

  return (
    <>
      <Direction>
        <StyledOperationTable>
          <OrderBy options={ORDERBY_OPTIONS} />

          <SortBy options={ROOM_PAGE.SORTBY_OPTIONS} />
          <Search setPlaceHolder="Search by name..." />
        </StyledOperationTable>

        {isLoading && <Spinner />}

        {rooms && rooms.length ? (
          <Menus>
            <Table columns="10% 40% 20% 20% 5%">
              <Table.Header headerColumn={columnName} />
              <Table.Body<IRoom> data={rooms} render={renderRoomRow} />
              <Table.Footer>
                <Pagination count={count!}/>
              </Table.Footer>
            </Table>
          </Menus>
        ) : (
          !isLoading && <Message>No data to show here!</Message>
        )}
      </Direction>
    </>
  );
};

export default RoomTable;
