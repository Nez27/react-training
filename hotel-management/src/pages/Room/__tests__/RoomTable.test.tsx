import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { BrowserRouter } from 'react-router-dom';
import { render } from '@testing-library/react';

// Components
import RoomTable from '@page/Room/RoomTable';

describe('UserTable', () => {
  const queryClient = new QueryClient();

  const wrapper = render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>
        <RoomTable />
      </BrowserRouter>
    </QueryClientProvider>
  );

  test('Should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
