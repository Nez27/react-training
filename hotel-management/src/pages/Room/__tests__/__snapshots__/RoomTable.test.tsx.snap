// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTable Should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-giAruI eEOXtt"
        type="vertical"
      >
        <div
          class="sc-iJCSeZ boUFen"
        >
          <div
            class="sc-dItHI exgtuU"
          >
            <button
              class="sc-bqGGcD entqcN"
              disabled=""
            >
              Ascending
            </button>
            <button
              class="sc-bqGGcD graLyf"
            >
              Descending
            </button>
          </div>
          <select
            aria-label="Sort"
            class="sc-crzpnZ iXFhDe"
          >
            <option
              value="id"
            >
              Sort by id
            </option>
            <option
              value="name"
            >
              Sort by name
            </option>
            <option
              value="price"
            >
              Sort by price
            </option>
          </select>
          <input
            class="sc-iqAbyq kOrhXJ"
            placeholder="Search by name..."
          />
        </div>
        <div
          class="sc-ezyZrH gLRBxl"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-giAruI eEOXtt"
      type="vertical"
    >
      <div
        class="sc-iJCSeZ boUFen"
      >
        <div
          class="sc-dItHI exgtuU"
        >
          <button
            class="sc-bqGGcD entqcN"
            disabled=""
          >
            Ascending
          </button>
          <button
            class="sc-bqGGcD graLyf"
          >
            Descending
          </button>
        </div>
        <select
          aria-label="Sort"
          class="sc-crzpnZ iXFhDe"
        >
          <option
            value="id"
          >
            Sort by id
          </option>
          <option
            value="name"
          >
            Sort by name
          </option>
          <option
            value="price"
          >
            Sort by price
          </option>
        </select>
        <input
          class="sc-iqAbyq kOrhXJ"
          placeholder="Search by name..."
        />
      </div>
      <div
        class="sc-ezyZrH gLRBxl"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
